File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
       
        self.result = -float("inf")

        def find_max(node):
            if node is None:
                return 0
            
            left_max = max(find_max(node.left), 0)
            right_max = max(find_max(node.right), 0)

            self.result = max(self.result, left_max + node.val + right_max)

            return max(left_max + node.val, right_max + node.val)

        find_max(root)

        return self.result

        

         

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int maxPath = INT_MIN;
    int getMaxPath(TreeNode *root) {
        if (root == nullptr) {
            return 0;
        }

        auto left = max(getMaxPath(root->left), 0);
        auto right = max(getMaxPath(root->right), 0);

        auto fullPath = root->val + left + right;

        maxPath = max(maxPath, fullPath);

        return max(root->val + left, root->val + right);
    }
public:
    int maxPathSum(TreeNode* root) {
        getMaxPath(root);
        return maxPath;
    }
};
